### ISSUE TITLE
Support running tests with limited resources (e.g. browser based)  

### ISSUE THREAD

**Issue by pawelpabich** [xunit/xunit#107]
Hi, 

I've been looking for a while for a test framework that would let me run browser based tests in parallel and it looks like xUnit has most of the required features. I blogged a detailed analysis here ([http://www.pabich.eu/2014/06/how-to-run-selenium-tests-in-parallel.html](http://www.pabich.eu/2014/06/how-to-run-selenium-tests-in-parallel.html)) but I will include all important information here so the discussion is self-contained. The purpose of this discussion is to see whether those missing features fit into the vision of the xUnit project and if so, how they can be implemented. Personally I believe that the famous extensibility capabilites of xUnit could help here a lot (e.g. custom test execution pipeline via a custom implementation of XUnitTestCase). And having one, instead of many frameworks, has its obvious adventages.   

I forked xUnit [https://github.com/pawelpabich/xunit](https://github.com/pawelpabich/xunit) to see if I can fill the gaps. This was easier than I thought it would be thanks to great condition the xUnit codebase is in. The whole excercise is more a Proof of Concept then a fully blown implementation but it works and it is used successfully in production.

I split the missing features into 3 groups: 
- must have (without them browser based tests won't be possible with xUnit),
- should have (very,very usefull and it will make life much easier but not deal breakers)
- nice to have (there is no rush to add those, but eventually it would good to have them)
### Report test result back to the test before it gets disposed

Tests need to know when they fail so a screenshot of the web browser can be taken. [Current solution checks whether the test class implements certain interfaces, e.g. INeedToKnowTestFailure](https://github.com/pawelpabich/xunit/commit/77d79aa316dc57950c9f5c110dc72f25111f7127). 
### Execute the whole test as a single Task

Current design splits the execution pipline into mutliple tasks which means that Dispose method of an already executed test might be run with some delay which means more browser instances will be required than what MaxParallelThreads would suggest. In my tests I very often ended up with 2 or 3 x MaxParallelThreads number of browsers. [Current solution is to execute the whole pipleline as a single Task](https://github.com/pawelpabich/xunit/commit/7fa0f32a1f831f8aef55a8ff7ee6db0dfb88a8cd)
### Make unit of work more granular

At the moment the smalles unit of work for parallel processing is class. For long running tests this is not optimal as a class with a few long running tests can significantly slow down the whole build. [Current solution introduces a new, per method, collection bahaviour.](https://github.com/pawelpabich/xunit/commit/7fa0f32a1f831f8aef55a8ff7ee6db0dfb88a8cd)
### Allow to order test collections

Assuming that we can have a test collection per test methed then the next logical step to minimize the exectuion time would be to have ability to order test collections to make sure the failed tests or/and tests with longest execution time are run first. [Current solution is based on the existing extensibilty point for ordering tests within a test collection.](https://github.com/pawelpabich/xunit/commit/74eb583589f01711137a754d7b90e5b7a79bb1e4)
### Allow to perform global initalization

At the moment global [initialization needs to happen in the instance constructor with global lock.](https://github.com/pawelpabich/SeleniumWithXunit/blob/master/SeleniumWithXunit/Infrastructure/BaseTest.cs?#L48-58)
### Allow to perform global clean up

Current workaround relies on [subscription to AppDomain.Unload](https://github.com/pawelpabich/SeleniumWithXunit/blob/master/SeleniumWithXunit/Infrastructure/TestRun.cs?#L16) event which works most of the time but not always.

I really hope we can make it work :). Thoughts?

**Comment by jfloodnet** [comment]
+1

**Comment by RichiCoder1** [comment]
:+1: Just running into this. Deterministic disposal and any easy way to globally setup and teardown state are my biggest issues.

**Comment by bradwilson** [comment]
The demands at the top of this issue are not things we are planning to deal with; our focus with xUnit.net is to keep the core focused on TDD, with extensibility to let other people adapt it to their needs. Our recommendation is still to write your own test framework (implementation of `ITestFramework`). There is an example of this here: https://github.com/xunit/samples.xunit/tree/master/ObservationExample

**Comment by RichiCoder1** [comment]
That's how I ended up solving it. Abstracted away TestFramework and TestExecutor to handle IIS Express spin up. If anyone else in this thread is interested in building a general purpose selenium test framework on top of xUnit using the xUnit.Sdk, let me know.

**Comment by pawelpabich** [comment]
@RichiCoder1 would love to have a look

**Comment by jfloodnet** [comment]
@pawelpabich @RichiCoder1 me too - at this stage I've just subclassed the XUnitTestCaseRunner.

**Comment by RichiCoder1** [comment]
Reading over the goals you were aiming for @pawelpabich, I think we had slightly different use scenarios. I was most interested in per-assembly Setup and Teardown, and I prefer the class-as-collection and per-collection behavior. That being said, I'm cleaning up and de-companyizing my code and will push it out to GitHub in a little. After that, we can make a few issues and see about building an all-encompassing, nuget-pluggable solution. Sound good?

**Comment by pawelpabich** [comment]
Sounds like a good start @RichiCoder1 :)

**Comment by RichiCoder1** [comment]
Dropped a *very* basic framework here: https://github.com/RichiCoder1/Browser.xUnit

Going to add the custom setup and teardown extensibility points I added elsewhere here in a little. Feel free to create issues and start discussions. My goals are very simplistic and I want to make sure we come up with a good solution :)

**Comment by pawelpabich** [comment]
@RichiCoder1 looks good, thanks!

**Comment by bradwilson** [comment]
Closing this, as it's something that should be done externally.
